File: app/src/main/java/com/mitsuki/mosterhunterworldwiki/base/BasePresenter.java
package com.mitsuki.mosterhunterworldwiki.base;

import android.content.Context;

public abstract class BasePresenter<V extends IView, M extends IModel> {
    protected V mView;
    protected M mModel;

    protected Context mContext;

    public BasePresenter(Context context, V mView) {
        this.mView = mView;
        this.mModel = getModel();
        this.mContext = context;
    }

    public void onDestroy() {
        if (mModel != null)
            mModel.onDestroy();
    }

    protected abstract M getModel();
}

File: simplemvp/src/main/java/com/mitsuki/simplemvp/base/BasePresenter.kt
package com.mitsuki.simplemvp.base

import org.kodein.di.Kodein

abstract class BasePresenter<V : IView, M : IModel>(kodein: Kodein) {
    abstract val mView: V
    open val mModel: M? = null

    fun onDestroy() {
        mModel?.onDestroy()
    }
}
